pinot-servicemanager/src/main/java/org/hypertrace/pinot/plugins/GenericAvroMessageDecoder.java

```diff
@@ -0,0 +1,65 @@
+package org.hypertrace.pinot.plugins;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericData.Record;
+import org.apache.pinot.plugin.inputformat.avro.AvroRecordExtractor;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.data.readers.RecordExtractor;
+import org.apache.pinot.spi.plugin.PluginManager;
+import org.apache.pinot.spi.stream.StreamMessageDecoder;
+import org.hypertrace.core.kafkastreams.framework.serdes.GenericAvroSerde;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@NotThreadSafe
+public class GenericAvroMessageDecoder implements StreamMessageDecoder<byte[]> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(GenericAvroMessageDecoder.class);
+  GenericAvroSerde genericAvroSerde;
+  String topicName;
+  private RecordExtractor<Record> _avroRecordExtractor;
+
+
+  @Override
+  public void init(Map<String, String> props, Set<String> fieldsToRead, String topicName)
+      throws Exception {
+    this.topicName = topicName;
+
+    genericAvroSerde = new GenericAvroSerde();
+    genericAvroSerde.configure(props, false);
+
+    String recordExtractorClass = props.get(RECORD_EXTRACTOR_CONFIG_KEY);
+    // Backward compatibility to support Avro by default
+    if (recordExtractorClass == null) {
+      recordExtractorClass = AvroRecordExtractor.class.getName();
+    }
+    _avroRecordExtractor = PluginManager.get().createInstance(recordExtractorClass);
+    _avroRecordExtractor.init(fieldsToRead, null);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>NOTE: the payload should contain message content only (without header).
+   */
+  @Override
+  public GenericRow decode(byte[] payload, GenericRow destination) {
```

Comment:
So as far as I understand this will read the payload from the message and ignore the schema ID to proceed with a local avro schema?

Reply:
write schema will be part of the payload if we use `genericAvroSerde`. So, internally, it will first read the schema and use it. Can check here - https://github.com/hypertrace/kafka-streams-framework/blob/main/kafka-streams-serdes/src/main/java/org/hypertrace/core/kafkastreams/framework/serdes/GenericAvroDeserializer.java